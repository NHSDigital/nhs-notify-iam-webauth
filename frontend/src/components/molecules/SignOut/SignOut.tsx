'use client';

import { useEffect } from 'react';
import { useAuthenticator } from '@aws-amplify/ui-react';
import { signOut } from '@aws-amplify/auth';
import JsCookie from 'js-cookie';
import { authenticatorSelector } from '@/src/utils/authenticator-selector';

export const SignOut = ({ children }: { children?: React.ReactNode }) => {
  const { authStatus } = useAuthenticator(authenticatorSelector);

  useEffect(() => {
    if (authStatus === 'authenticated') {
      signOut({ global: true });
      JsCookie.remove('csrf_token');
    }
  }, [authStatus]);

  return authStatus === 'authenticated' ? <p>Signing out</p> : children;
};
