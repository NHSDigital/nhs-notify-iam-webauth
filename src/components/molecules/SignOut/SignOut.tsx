'use client';

import { useEffect } from 'react';
import { useAuthenticator } from '@aws-amplify/ui-react';
import JsCookie from 'js-cookie';
import { authenticatorSelector } from '@/src/utils/authenticator-selector';

export const SignOut = ({ children }: { children?: React.ReactNode }) => {
  const { signOut, authStatus } = useAuthenticator(authenticatorSelector);

  useEffect(() => {
    if (authStatus === 'authenticated') {
      signOut({ global: true });
      JsCookie.remove('csrf_token');
    }
  }, [authStatus, signOut]);

  return authStatus === 'authenticated' ? <p>Signing out</p> : children;
};
