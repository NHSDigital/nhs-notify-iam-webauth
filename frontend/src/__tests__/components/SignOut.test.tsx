import { mockDeep } from 'jest-mock-extended';
import { render } from '@testing-library/react';
import { useAuthenticator, UseAuthenticator } from '@aws-amplify/ui-react';
import JsCookie from 'js-cookie';
import { SignOut } from '@/src/components/molecules/SignOut/SignOut';

const signOutMock = jest.fn();
jest.mock('@aws-amplify/ui-react', () => ({
  useAuthenticator: jest.fn(() =>
    mockDeep<UseAuthenticator>({
      authStatus: 'authenticated',
      signOut: signOutMock,
    })
  ),
}));
jest.mock('js-cookie');

const useAuthenticatorMock = jest.mocked(useAuthenticator);
const cookieMock = jest.mocked(JsCookie);

describe('Signout', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('calls signOut when user is authenticated', async () => {
    const page = render(
      <SignOut>
        <p>signed out</p>
      </SignOut>
    );

    expect(signOutMock).toHaveBeenCalled();
    expect(cookieMock.remove).toHaveBeenCalledWith('csrf_token');

    expect(page.asFragment()).toMatchSnapshot();
  });

  test('does not call signOut when user is unauthenticated', async () => {
    useAuthenticatorMock.mockReturnValueOnce(
      mockDeep<UseAuthenticator>({ authStatus: 'unauthenticated' })
    );
    const page = render(
      <SignOut>
        <p>signed out</p>
      </SignOut>
    );

    expect(signOutMock).not.toHaveBeenCalled();
    expect(cookieMock.remove).not.toHaveBeenCalled();

    expect(page.asFragment()).toMatchSnapshot();
  });
});
