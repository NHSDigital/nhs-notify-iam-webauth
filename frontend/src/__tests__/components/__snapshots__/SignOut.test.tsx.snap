// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signout calls signOut when user is authenticated 1`] = `
<DocumentFragment>
  <p>
    Signing out
  </p>
</DocumentFragment>
`;

exports[`Signout does not call signOut when user is unauthenticated 1`] = `
<DocumentFragment>
  <p>
    signed out
  </p>
</DocumentFragment>
`;
